Refactor/subscriber counts into redux store

chrismccluskey opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Closes 88-gh-Automattic/gold

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Fetches subscriber counts from data store instead of passing it to the child component as props from parents.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* On an installation of WordPress, make sure the Jetpack Beta plugin is installed, and then activate this branch.
* Check this patch out to sun on sandbox.
* Ensure subscribe and public-api are pointed to sandbox.
* Go to [the earn settings page](https://wordpress.com/earn/payments/) and ensure Stripe is setup in test mode.
* Create a new page or post and add the subscribe block.
* Publish the page and try subscribing. Confirm the block works as expected.
* Return to edit the page with the subscription button. Click on the "Newsletter Visibility" drop-down and select "paid subscribers". You should not see a message indicating that you don't have any subscribers. Instead, you should see a message saying this will be sent to your _n_ subscribers.

chrismccluskey committed:
Store social counts in memberships-product store

chrismccluskey committed:
changelog

chrismccluskey committed:
Fix issue with build

millerf commented on projects/plugins/jetpack/extensions/store/membership-products/reducer.js:
```diff
@@ -23,6 +30,34 @@ export default function reducer( state = DEFAULT_STATE, action ) {
+				...state,
+				counts: {
+					...state.counts,
+					paidSubscribers: action.socialFollowers,
```
Might be an error there
```suggestion
					paidSubscribers: action.paidSubscribers,
```

millerf commented on projects/plugins/jetpack/extensions/store/membership-products/reducer.js:
```diff
@@ -23,6 +30,34 @@ export default function reducer( state = DEFAULT_STATE, action ) {
+				...state,
+				counts: {
+					...state.counts,
+					emailSubscribers: action.socialFollowers,
```
might be an error there

```suggestion
					emailSubscribers: action.emailSubscribers,
```

chrismccluskey committed:
Fix typo

chrismccluskey committed:
Fix casing and remove commented code

chrismccluskey committed:
Transition accessLevel and showMisconfigurationWarning to store

chrismccluskey committed:
Fix incorrect usage of hooks

millerf commented on projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js:
```diff
@@ -115,18 +99,16 @@ const NewsletterDisabledPanels = () => (
+function NewsletterPrePublishSettingsPanel( { setPostMeta, isModuleActive, showPreviewModal } ) {
 	const { tracks } = useAnalytics();
 	const { changeStatus, isLoadingModules, isChangingStatus } = useModuleStatus( name );
+	const { accessLevel } = useSelect( select => {
```
Wouldn't writing it this way be a bit more readable? 

```suggestion
	const accessLevel = useSelect( select =>  select( 'jetpack/membership-products').getAccessLevel() );
```

Instead of wrapping it in an object to extract it straight away?

chrismccluskey replied:
Yes, this is much more readable. I'll make these changes, thanks!

millerf replied:
[Object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) is pretty useful and powerful in JS

millerf commented on projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js:
```diff
@@ -194,12 +169,20 @@ function NewsletterPrePublishSettingsPanel( {
+			accessLevel: getAccessLevel(),
+			showMisconfigurationWarning: getShowMisconfigurationWarning(),
+		};
+	} );
```
Same comment than previously...

I would write it:

```
const accessLevel = useSelect( select =>  select( 'jetpack/membership-products').getAccessLevel() );
const showMisconfigurationWarning = useSelect( select =>  select( 'jetpack/membership-products').getShowMisconfigurationWarning() );
```

millerf commented on projects/plugins/jetpack/extensions/store/membership-products/resolvers.js:
```diff
@@ -61,6 +73,41 @@ const mapAPIResponseToMembershipProductsStoreData = ( response, registry, dispat
+	registry,
+	dispatch
+) => {
+	dispatch( setSocialFollowerCount( response.counts.social_followers ) );
```
Does each dispatch() triggers a new component refresh?

chrismccluskey replied:
I'm not sure if it does. I'll need to check this and see. From memory, there were no visuals which might indicate a refresh, although it's possible one still may have happened.

millerf replied:
It is fine. Let's keep it that way...

chrismccluskey committed:
Make syntax more readable and start reverting accessLevel changes

chrismccluskey committed:
Revert more of accessLevel changes

chrismccluskey committed:
Revert showMisconfigurationWarning changes

chrismccluskey committed:
Use components correctly

chrismccluskey committed:
Add missing param and remove action

chrismccluskey committed:
Remove empty line

chrismccluskey committed:
Merge remote-tracking branch 'origin/trunk' into refactor/subscriber-counts-into-redux-store

millerf committed:
Add a getShowMisconfigurationWarning utility method.

millerf committed:
define and use setPostMeta only when necessary

millerf committed:
[not verified] Revert "define and use setPostMeta only when necessary"

This reverts commit 278a6c0300769ec41e6e2b101487f337b96b094e.

millerf commented on projects/plugins/jetpack/extensions/blocks/subscriptions/panel.js:
```diff
@@ -323,12 +314,8 @@ export default function SubscribePanels() {
-			setPaidSubscribers( counts.paid_subscribers );
-		} );
-	}, [ isModuleActive ] );
+		getSubscriberCounts();
```
That should not be needed as it should be handled by resolvers?

millerf committed:
Prevent having to retrieve the subscribers out of context

- Remove getSubscriberCounts() call in subscribePanels
- Set only one event in the store for the full subscriber count retrieval

millerf committed:
Clean up code

millerf reviewed (approved):
LGTM. The code looks a bit more clean!

I added a couple of commits:
- e7e1842f16 to move getShowMisconfigurationWarning to its own utility fucntion
- 278a6c0300 I tried to also remove setPostMeta, but on changing the panel is not refreshed (it would need some extra work to pass it in the data store), so I reverted it. Not needed really.
- c6e3f496ee to prevent calling getSubscriberCounts() out of context. There was no need to have an event for each of the separate counts. Now it make things easier to read.

Nice work anyway for the first React PR in a while.

jeherve commented:
This is introducing a regression. When the Subscriptions module is not active, the `wpcom/v2/subscribers/counts` is not available so we should not attempt to make a request. Before this PR, we waited until the module was active to make a request ; now we get a 404 as soon as we load the post editor:

<img width="831" alt="image" src="https://github.com/Automattic/jetpack/assets/426388/ceac1e85-78ba-433e-9783-8e556aac8e55">

I assume that eventually we'll get rid of `api.js` entirely, to avoid the duplication?

millerf commented:
> This is introducing a regression. When the Subscriptions module is not active, the `wpcom/v2/subscribers/counts` is not available so we should not attempt to make a request. Before this PR, we waited until the module was active to make a request ; now we get a 404 as soon as we load the post editor:
> 
> <img alt="image" width="831" src="https://user-images.githubusercontent.com/426388/258111693-ceac1e85-78ba-433e-9783-8e556aac8e55.png">
> I assume that eventually we'll get rid of `api.js` entirely, to avoid the duplication?

Interesting. I am not sure why the data is called... AFAIK it should not get triggered... I'll look into it.
Thanks for finding this out..

millerf commented:
> I assume that eventually we'll get rid of api.js entirely, to avoid the duplication?

Yes that is a good idea :)
